tests/importer/importer.cpp

```diff
@@ -0,0 +1,761 @@
+    std::string funkyString = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+                              "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"funky\">\n"
+                              // Status quo functionality, key is local URL + base file path as the local URL doesn't exist as a key.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"units_imported.cellml\">\n"
+                              "    <units units_ref=\"units1_imported\" name=\"units1FromLibrary\"/>\n"
+                              "  </import>\n"
+                              // Use the local URL for a hidden child source component.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"components_source.cellml\">\n"
+                              "    <component component_ref=\"component1\" name=\"component1FromLibrary\"/>\n"
+                              "  </import>\n"
+                              // Use the absolute URL for a hidden child source component.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\""
+                              + resourcePath("importer/components_source.cellml")
+                              + "\">\n"
+                                "    <component component_ref=\"component2\" name=\"component2FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Remote website URL used for key.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"https://www.example.com/myComponents.cellml\">\n"
+                                "    <component component_ref=\"component3\" name=\"component3FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Currently non-existent file used for the key. For this CellML file to function outside of the
+                                // the Importer instance, this model needs to be written to this location.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"i_dont_exist_yet.cellml\">\n"
+                                "    <units units_ref=\"units4\" name=\"units4FromLibrary\"/>\n"
+                                "  </import>\n"
+                                "</model>";
+
+    auto funkyModel = parser->parseModel(funkyString);
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    // Pass in a base file path, this will be used to resolve URLs if they don't already exist as keys.
+    // This allows the model to be resolved against different locations where local files are specified.
+    importer->resolveImports(funkyModel, resourcePath("importer/"));
+    EXPECT_FALSE(funkyModel->hasUnresolvedImports());
+    EXPECT_EQ(size_t(0), importer->issueCount());
+
+    // Apparently funky hrefs are valid ...
```

Comment:
Rephrase that comment. It looks like we are surprised by our API's behaviour. We shouldn't be... normally. :)

Reply:
b6b60a6d